src/teamspeak/mod.rs

```diff
@@ -86,50 +85,82 @@ fn get_message(event: &Event) -> Option<MusicBotMessage> {
+            while let Some(item) = conn.next().await {
+                use SyncStreamItem::*;
+
+                match item {
+                    Ok(ConEvents(events)) => {
                         for event in &events {
                             if let Some(msg) = get_message(event) {
-                                let tx = tx.read().expect("RwLock was not poisoned");
-                                // Ignore the result because the receiver might get dropped first.
-                                let _ = tx.send(msg);
+                                tokio::spawn(bot.send(msg));
                             }
                         }
                     }
-                    Err(e) => error!("Error occured during event reading: {}", e),
-                    Ok(SyncStreamItem::DisconnectedTemporarily) => debug!("Temporary disconnect!"),
-                    _ => (),
+                    Err(e) => error!(ev_logger, "Error occured during event reading: {}", e),
+                    Ok(DisconnectedTemporarily(r)) => {
+                        debug!(ev_logger, "Temporary disconnect"; "reason" => ?r)
+                    }
+                    Ok(Audio(_)) => {
+                        trace!(ev_logger, "Audio received");
+                    }
+                    Ok(IdentityLevelIncreasing(_)) => {
+                        trace!(ev_logger, "Identity level increasing");
+                    }
+                    Ok(IdentityLevelIncreased) => {
+                        trace!(ev_logger, "Identity level increased");
+                    }
+                    Ok(NetworkStatsUpdated) => {
+                        trace!(ev_logger, "Network stats updated");
+                    }
                 }
             }
-        }));
-
-        handle.wait_until_connected().await?;
-
-        let mut chandle = handle.clone();
-        chandle
-            .with_connection(|mut conn| {
-                conn.get_state()
-                    .expect("is connected")
-                    .server
-                    .set_subscribed(true)
-                    .send(&mut conn)
-                    .unwrap()
-            })
-            .await
-            .unwrap();
-
-        Ok(TeamSpeakConnection { handle })
+        });
+
+        let mut handle = self.handle.clone();
+        tokio::spawn(async move {
+            handle
+                .as_mut()
+                .expect("connect_for_bot was called")
+                .wait_until_connected()
+                .await
+                .unwrap();
+            handle
+                .as_mut()
+                .expect("connect_for_bot was called")
+                .with_connection(|mut conn| {
+                    conn.get_state()
+                        .expect("can get state")
+                        .server
+                        .set_subscribed(true)
+                        .send(&mut conn)
+                })
+                .await
+                .and_then(|v| v)
```

Comment:
isn't this a no-op?

Reply:
it flattens `Result<Result<T, E>, E>`

Reply:
std will do the same: https://doc.rust-lang.org/src/core/result.rs.html#1211